packages/driver/src/cy/snapshots.ts

```diff
@@ -10,6 +10,94 @@ export const HIGHLIGHT_ATTR = 'data-cypress-el'
 
 export const FINAL_SNAPSHOT_NAME = 'final state'
 
+type SelectorNode = {
+  frameId?: string
+  selector: string
+  ownerDoc: Document | ShadowRoot
+  host?: SelectorNode
+}
+
+const returnShadowRootIfShadowDomNode = (node: Element): ShadowRoot | null => {
+  // The method for checking if a node is a shadowRoot is a bit wonky, but seems to be the most sound.
+  // the shadowRoot object property only lives on the node context OUTSIDE the shadow DOM, meaning that
+  // node.parentNode.host.shadowRoot would work. But this is considered an instance of an Object and not
+  // a ShadowRoot, so likely checking the string serialization might be a safer choice here.
+  const isNodeShadowRoot = (n: any) => n?.toString() === '[object ShadowRoot]'
+
+  let parent = node && node.parentNode
+
+  while (parent) {
+    if (isNodeShadowRoot(parent)) {
+      return parent as ShadowRoot
+    }
+
+    parent = parent.parentNode
+  }
+
+  return null
+}
+
+/**
+ *
+ * @param elem - either an HTML Element or an element that lives within the ShadowDom or the Root Document
+ * @returns SelectorNode if the selector can be discovered. For regular elements, this should only be one object deep, but for ShadowDom
+ * elements, the SelectorNode tree could be N levels deep until the root is discovered
+ */
+function constructElementSelectorTree (elem: Element): SelectorNode | undefined {
+  try {
+    const ownerDoc = elem.ownerDocument
+    const elWindow = ownerDoc.defaultView
+
+    if (elWindow === null) {
+      return undefined
+    }
+
+    // finder tries to find the shortest unique selector to an element,
+    // but since we are more concerned with speed, we set the threshold to 1 and maxNumberOfTries to 0
+    // @ts-expect-error because 'root' can be either Document or Element but is defined as Element
+    // @see https://github.com/antonmedv/finder/issues/75
+    const selector = finder(elem, { root: ownerDoc, threshold: 1, maxNumberOfTries: 0 })
+
+    if (!selector) {
```

Comment:
Does `finder` return an `undefined`/`null` selector?

Reply:
If I remember correctly `finder` returns `''` as the selector

Reply:
Does this code need to be removed or modified then since this is just checking for the existence?

Reply:
Looking at the `finder` code, I'm not seeing where an empty string selector would be returned. If the selector is not found, an error is thrown. Am I missing something?

Reply:
Also took a look and that behavior looks correct, so I think this just needs to be removed. I removed it in 53b9b247ef